### ISSUE TITLE
Collection upload broken

### ISSUE THREAD

**Issue by dbk-rabel** [redhat-cop/ah_configuration#380]
Seems not to be a duplicate of #292 because of a different error message. Not 100% sure though.

# Summary

Collection upload does not seem to work anymore.

# Issue Type

- Bug Report

# Ansible, Collection, Private Automation Hub details

```console (paste below)
ansible --version
ansible [core 2.15.3]
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/home/automation/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3.11/site-packages/ansible
  ansible collection location = /home/automation/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.11.5 (main, Sep 22 2023, 15:34:29) [GCC 8.5.0 20210514 (Red Hat 8.5.0-20)] (/usr/bin/python3.11)
  jinja version = 3.1.2
  libyaml = True


ansible-galaxy collection list

# /home/automation/.ansible/collections/ansible_collections
Collection                    Version
----------------------------- -------
ansible.controller            4.5.2
ansible.posix                 1.5.4
infra.aap_utilities           2.3.0
infra.ah_configuration        2.0.6
kubernetes.core               3.0.0

# /usr/lib/python3.11/site-packages/ansible_collections
Collection                    Version
----------------------------- -------
amazon.aws                    6.3.0
ansible.netcommon             5.1.2
ansible.posix                 1.5.4
ansible.utils                 2.10.3
ansible.windows               1.14.0
arista.eos                    6.0.1
awx.awx                       22.6.0
azure.azcollection            1.16.0
check_point.mgmt              5.1.1
chocolatey.chocolatey         1.5.1
cisco.aci                     2.7.0
cisco.asa                     4.0.1
cisco.dnac                    6.7.3
cisco.intersight              1.0.27
cisco.ios                     4.6.1
cisco.iosxr                   5.0.3
cisco.ise                     2.5.14
cisco.meraki                  2.15.3
cisco.mso                     2.5.0
cisco.nso                     1.0.3
cisco.nxos                    4.4.0
cisco.ucs                     1.10.0
cloud.common                  2.1.4
cloudscale_ch.cloud           2.3.1
community.aws                 6.2.0
community.azure               2.0.0
community.ciscosmb            1.0.6
community.crypto              2.15.0
community.digitalocean        1.24.0
community.dns                 2.6.0
community.docker              3.4.8
community.fortios             1.0.0
community.general             7.3.0
community.google              1.0.0
community.grafana             1.5.4
community.hashi_vault         5.0.0
community.hrobot              1.8.1
community.libvirt             1.2.0
community.mongodb             1.6.1
community.mysql               3.7.2
community.network             5.0.0
community.okd                 2.3.0
community.postgresql          2.4.3
community.proxysql            1.5.1
community.rabbitmq            1.2.3
community.routeros            2.9.0
community.sap                 1.0.0
community.sap_libs            1.4.1
community.skydive             1.0.0
community.sops                1.6.4
community.vmware              3.9.0
community.windows             1.13.0
community.zabbix              2.1.0
containers.podman             1.10.2
cyberark.conjur               1.2.0
cyberark.pas                  1.0.19
dellemc.enterprise_sonic      2.2.0
dellemc.openmanage            7.6.1
dellemc.powerflex             1.7.0
dellemc.unity                 1.7.1
f5networks.f5_modules         1.25.1
fortinet.fortimanager         2.2.1
fortinet.fortios              2.3.1
frr.frr                       2.0.2
gluster.gluster               1.0.2
google.cloud                  1.2.0
grafana.grafana               2.1.5
hetzner.hcloud                1.16.0
hpe.nimble                    1.1.4
ibm.qradar                    2.1.0
ibm.spectrum_virtualize       1.12.0
infinidat.infinibox           1.3.12
infoblox.nios_modules         1.5.0
inspur.ispim                  1.3.0
inspur.sm                     2.3.0
junipernetworks.junos         5.2.0
kubernetes.core               2.4.0
lowlydba.sqlserver            2.1.0
microsoft.ad                  1.3.0
netapp.aws                    21.7.0
netapp.azure                  21.10.0
netapp.cloudmanager           21.22.0
netapp.elementsw              21.7.0
netapp.ontap                  22.7.0
netapp.storagegrid            21.11.1
netapp.um_info                21.8.0
netapp_eseries.santricity     1.4.0
netbox.netbox                 3.13.0
ngine_io.cloudstack           2.3.0
ngine_io.exoscale             1.0.0
ngine_io.vultr                1.1.3
openstack.cloud               2.1.0
openvswitch.openvswitch       2.1.1
ovirt.ovirt                   3.1.2
purestorage.flasharray        1.20.0
purestorage.flashblade        1.12.1
purestorage.fusion            1.6.0
sensu.sensu_go                1.14.0
servicenow.servicenow         1.0.6
splunk.es                     2.1.0
t_systems_mms.icinga_director 1.33.1
telekom_mms.icinga_director   1.34.1
theforeman.foreman            3.12.0
vmware.vmware_rest            2.3.1
vultr.cloud                   1.8.0
vyos.vyos                     4.1.0
wti.remote                    1.0.5



Private Automation Hub version
4.9.1
```

# OS / ENVIRONMENT

RHEL8.9

# Desired Behavior

Collection uploads via infra.ah_configuration.collection works.

# Actual Behavior

```console (Ansible-Output)
TASK [infra.ah_configuration.namespace : Create Namespace | Wait for finish the namespace creation] *****************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************
FAILED - RETRYING: [vm-9edd65a2c92c.it.debeka.test]: Create Namespace | Wait for finish the namespace creation (600 retries left).
changed: [vm-9edd65a2c92c.it.debeka.test] => (item={'failed': 0, 'started': 1, 'finished': 0, 'ansible_job_id': 'j634380225762.976909', 'results_file': '/home/automation/.ansible_async/j634380225762.976909', 'changed': False, '__namespace_item': {'name': 'debeka_automation', 'company': 'Debeka'}, 'ansible_loop_var': '__namespace_item'})
changed: [vm-9edd65a2c92c.it.debeka.test] => (item={'failed': 0, 'started': 1, 'finished': 0, 'ansible_job_id': 'j102679503520.976930', 'results_file': '/home/automation/.ansible_async/j102679503520.976930', 'changed': False, '__namespace_item': {'name': 'community'}, 'ansible_loop_var': '__namespace_item'})

TASK [Include role infra.ah_configuration.collection] ***************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************

TASK [infra.ah_configuration.collection : Validating arguments against arg spec 'main' - An Ansible Role to update, or destroy Automation Hub Collections.] *********************************************************************************************************************************************************************************************************************************************************************************************************************************
ok: [vm-9edd65a2c92c.it.debeka.test]

TASK [infra.ah_configuration.collection : Update or destroy Automation Hub Collection] ******************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************
ok: [vm-9edd65a2c92c.it.debeka.test] => (item={'namespace': 'debeka_automation', 'name': 'infrastructure', 'version': '1.9.0', 'path': '/home/automation/aap_bundle_working_dir_site2/debeka_automation-infrastructure-1.9.0.tar.gz', 'state': 'present'})
ok: [vm-9edd65a2c92c.it.debeka.test] => (item={'namespace': 'community', 'name': 'general', 'version': '7.5.3', 'path': '/home/automation/aap_bundle_working_dir_site2/community-general-7.5.3.tar.gz', 'state': 'present'})
ok: [vm-9edd65a2c92c.it.debeka.test] => (item={'namespace': 'infra', 'name': 'controller_configuration', 'version': '2.7.1', 'path': '/home/automation/aap_bundle_working_dir_site2/infra-controller_configuration-2.7.1.tar.gz', 'state': 'present'})

TASK [infra.ah_configuration.collection : Sleep for 10 seconds and continue with play] ******************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************
ok: [vm-9edd65a2c92c.it.debeka.test]

TASK [infra.ah_configuration.collection : Update/Destroy collection | Wait for finish the update/destruction of collection] *****************************************************************************************************************************************************************************************************************************************************************************************************************************************************************
failed: [vm-9edd65a2c92c.it.debeka.test] (item={'failed': 0, 'started': 1, 'finished': 0, 'ansible_job_id': 'j761188713278.977011', 'results_file': '/home/automation/.ansible_async/j761188713278.977011', 'changed': False, '__collection': {'namespace': 'debeka_automation', 'name': 'infrastructure', 'version': '1.9.0', 'path': '/home/automation/aap_bundle_working_dir_site2/debeka_automation-infrastructure-1.9.0.tar.gz', 'state': 'present'}, 'ansible_loop_var': '__collection'}) => {"__collections_job_async_result_item": {"__collection": {"name": "infrastructure", "namespace": "debeka_automation", "path": "/home/automation/aap_bundle_working_dir_site2/debeka_automation-infrastructure-1.9.0.tar.gz", "state": "present", "version": "1.9.0"}, "ansible_job_id": "j761188713278.977011", "ansible_loop_var": "__collection", "changed": false, "failed": 0, "finished": 0, "results_file": "/home/automation/.ansible_async/j761188713278.977011", "started": 1}, "ansible_job_id": "j761188713278.977011", "ansible_loop_var": "__collections_job_async_result_item", "attempts": 1, "changed": false, "finished": 1, "msg": "You don't have permission to POST to api/galaxy/v3/artifacts/collections/ (HTTP 403).", "results_file": "/home/automation/.ansible_async/j761188713278.977011", "started": 1, "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
failed: [vm-9edd65a2c92c.it.debeka.test] (item={'failed': 0, 'started': 1, 'finished': 0, 'ansible_job_id': 'j844415255950.977032', 'results_file': '/home/automation/.ansible_async/j844415255950.977032', 'changed': False, '__collection': {'namespace': 'community', 'name': 'general', 'version': '7.5.3', 'path': '/home/automation/aap_bundle_working_dir_site2/community-general-7.5.3.tar.gz', 'state': 'present'}, 'ansible_loop_var': '__collection'}) => {"__collections_job_async_result_item": {"__collection": {"name": "general", "namespace": "community", "path": "/home/automation/aap_bundle_working_dir_site2/community-general-7.5.3.tar.gz", "state": "present", "version": "7.5.3"}, "ansible_job_id": "j844415255950.977032", "ansible_loop_var": "__collection", "changed": false, "failed": 0, "finished": 0, "results_file": "/home/automation/.ansible_async/j844415255950.977032", "started": 1}, "ansible_job_id": "j844415255950.977032", "ansible_loop_var": "__collections_job_async_result_item", "attempts": 1, "changed": false, "finished": 1, "msg": "You don't have permission to POST to api/galaxy/v3/artifacts/collections/ (HTTP 403).", "results_file": "/home/automation/.ansible_async/j844415255950.977032", "started": 1, "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}
ok: [vm-9edd65a2c92c.it.debeka.test] => (item={'failed': 0, 'started': 1, 'finished': 0, 'ansible_job_id': 'j866840532872.977052', 'results_file': '/home/automation/.ansible_async/j866840532872.977052', 'changed': False, '__collection': {'namespace': 'infra', 'name': 'controller_configuration', 'version': '2.7.1', 'path': '/home/automation/aap_bundle_working_dir_site2/infra-controller_configuration-2.7.1.tar.gz', 'state': 'present'}, 'ansible_loop_var': '__collection'})

PLAY RECAP **********************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************************
vm-9edd65a2c92c.it.debeka.test : ok=12   changed=2    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

```console (Pulpcore logs)
Apr 04 19:37:02 vm-d9978be7d402.it.debeka.test gunicorn[1747]: pulp [0620236c06674ad5b70d6563742a7d22]: pulpcore.tasking.tasks:INFO: Starting task 018eaa2f-7d5d-7d22-afea-805bbddabf1d
Apr 04 19:37:02 vm-d9978be7d402.it.debeka.test gunicorn[1747]: pulp [0620236c06674ad5b70d6563742a7d22]: pulpcore.tasking.tasks:INFO: Task completed 018eaa2f-7d5d-7d22-afea-805bbddabf1d
Apr 04 19:37:03 vm-d9978be7d402.it.debeka.test gunicorn[1723]: pulp [2bc3e15511a44ce3a1e1b6100b6b6434]: pulpcore.tasking.tasks:INFO: Starting task 018eaa2f-7ec8-7493-be58-81ede42b0a2f
Apr 04 19:37:03 vm-d9978be7d402.it.debeka.test gunicorn[1723]: pulp [2bc3e15511a44ce3a1e1b6100b6b6434]: pulpcore.tasking.tasks:INFO: Task completed 018eaa2f-7ec8-7493-be58-81ede42b0a2f
Apr 04 19:37:17 vm-d9978be7d402.it.debeka.test pulpcore-worker[15072]: pulp [b684cea42256401f81746d222e0aa478]: pulpcore.tasking.tasks:INFO: Starting task 018eaa2f-b4bd-739f-8c28-2f7eddd16ac6
Apr 04 19:37:17 vm-d9978be7d402.it.debeka.test pulpcore-worker[15072]: pulp [b684cea42256401f81746d222e0aa478]: pulpcore.tasking.tasks:INFO: Task completed 018eaa2f-b4bd-739f-8c28-2f7eddd16ac6
Apr 04 19:37:17 vm-d9978be7d402.it.debeka.test pulpcore-worker[15073]: pulp [21ec72663b3f411db58274c83334e6f2]: pulpcore.tasking.tasks:INFO: Starting task 018eaa2f-b5ba-7c4b-979b-8cd1f41439a1
Apr 04 19:37:17 vm-d9978be7d402.it.debeka.test pulpcore-worker[15073]: pulp [21ec72663b3f411db58274c83334e6f2]: pulpcore.tasking.tasks:INFO: Task completed 018eaa2f-b5ba-7c4b-979b-8cd1f41439a1
Apr 04 19:37:23 vm-d9978be7d402.it.debeka.test gunicorn[1723]: pulp [63b0e52cf8174bfcbb9c4510ef80bfa9]: django.request:WARNING: Not Found: /api/galaxy/v3/plugin/ansible/content/published/collections/index/debeka_automation/infrastructure/versions/1.9.0/
Apr 04 19:37:23 vm-d9978be7d402.it.debeka.test gunicorn[1741]: pulp [e321edc978644db1b2a26b0a869454f8]: django.request:WARNING: Forbidden: /api/galaxy/v3/artifacts/collections/
Apr 04 19:37:23 vm-d9978be7d402.it.debeka.test gunicorn[1751]: pulp [7af99e8d8ac743b7b4715c221f488bbf]: django.request:WARNING: Not Found: /api/galaxy/v3/plugin/ansible/content/published/collections/index/community/general/versions/7.5.3/
Apr 04 19:37:24 vm-d9978be7d402.it.debeka.test gunicorn[1747]: pulp [deefbd1466b447e9a48068508f2114dd]: django.request:WARNING: Forbidden: /api/galaxy/v3/artifacts/collections/
```

# STEPS TO REPRODUCE

```yaml (Important file from inventory)
ah_host: "{{ automationhub_main_url }}"
ah_username: "admin"
ah_password: "{{ aap_setup_prep_inv_vars['all']['automationhub_admin_password'] }}"
proxy_url: "{{ internal_http_proxy_url }}"


ah_configuration_async_retries: 600
ah_configuration_async_delay: 5

ah_collection_remotes:
  - name: "rh-certified"
    url: "https://console.redhat.com/api/automation-hub/content/published/"
    auth_url: "https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token"
    token: "{{ ansible_galaxy_token }}"
    signed_only: true
  - name: "validated"
    url: "https://console.redhat.com/api/automation-hub/content/validated/"
    auth_url: "https://sso.redhat.com/auth/realms/redhat-external/protocol/openid-connect/token"
    token: "{{ ansible_galaxy_token }}"
    signed_only: true
  - name: "community"
    url: "https://old-galaxy.ansible.com/"
    sync_dependencies: true
    signed_only: false
    requirements:
      - community.crypto
      - community.general
      - community.postgresql
      - community.rabbitmq
      - community.vmware
      - edb_devops.edb_postgres
      - jfrog.platform
      - netapp.ontap
      - theforeman.foreman
ah_collection_repositories:
  - name: "rh-certified"
    description: "Red Hat certified content repository"
    remote: "rh-certified"
    wait: true
    timeout: 3600
  - name: "validated"
    description: "Validated collections"
    remote: "validated"
    wait: true
    timeout: 3600
  - name: "community"
    description: "Community content repository"
    remote: "community"
    wait: true
    timeout: 3600
ah_namespaces:
  - name: "debeka_automation"
    company: "Debeka"
  - name: "community"
ah_collections:
  - namespace: "debeka_automation"
    name: "infrastructure"
    version: "1.9.0"
    path: "{{ aap_setup_prep_working_dir }}/debeka_automation-infrastructure-1.9.0.tar.gz"
    state: present
  - namespace: "community"
    name: "general"
    version: "7.5.3"
    path: "{{ aap_setup_prep_working_dir }}/community-general-7.5.3.tar.gz"
    state: present
  - namespace: "infra"
    name: "controller_configuration"
    version: "2.7.1"
    path: "{{ aap_setup_prep_working_dir }}/infra-controller_configuration-2.7.1.tar.gz"
    state: present

custom_collections_patterns:
  - "*.tar.gz"
```

```yaml (Playbook)
- name: Setup the AAP PAH configuration
  hosts:
    - bootstrap_node
  become: false
  gather_facts: false
  tasks:
    - name: Include tasks from role
      ansible.builtin.include_role:
        name: aap_install_tasks
        tasks_from: tasks_21_aap_hub_configure
``` 

```yaml (Tasks file)
---
- name: Setup the AAP PAH configuration block
  when: groups['pah_nodes']|length >0
  block:
    - name: Include role infra.ah_configuration.collection_remote
      ansible.builtin.include_role:
        name: infra.ah_configuration.collection_remote
    - name: Include role infra.ah_configuration.collection_repository
      ansible.builtin.include_role:
        name: infra.ah_configuration.collection_repository
    - name: Include role infra.ah_configuration.namespace
      ansible.builtin.include_role:
        name: infra.ah_configuration.namespace
    - name: Include role infra.ah_configuration.collection
      ansible.builtin.include_role:
        name: infra.ah_configuration.collection
```

Authentication seems to work in general, because repository and namespace task are not failing. Also the collection task does not fail, if the collection also exists.

The pulpcore logs suggest a problem with the upload.

Redhat support send me here, because the collection is not supported by Redhat.

Yours
  David

**Comment by Tompage1994** [comment]
Hi David,

Upon taking a look at your issue I can see you will likely have a problem trying to upload the `community.general` and `infra. controller_configuration` as they should be going into the community repository as set out in your requirements under `ah_collection_repositories`. This would therefore mean you shouldn't be trying to upload an artifact as I don't believe that will end well (I suspect you can't do this in the UI either).

With your `debeka_automation.infrastructure` collection this seems a slightly different issue. I think given the logs you've provided there isn't enough info to diagnose. I'd suggest taking a look in the tasks section of the UI when running this as this should return you a log of what happened during the attempted upload. Could you post this in please so we can diagnose further.

Thanks

**Comment by dbk-rabel** [comment]
Hi Tom.

Thanks for the quick answer!

> Upon taking a look at your issue I can see you will likely have a problem trying to upload the `community.general` and `infra. controller_configuration` as they should be going into the community repository as set out in your requirements under `ah_collection_repositories`. This would therefore mean you shouldn't be trying to upload an artifact as I don't believe that will end well (I suspect you can't do this in the UI either).

This is only true for community.general. I investigated that a little further:
1) Since we did not run a collection sync by now, the community.general collection is not in the PAH, only in the requirements.yml
2) I removed it from the requirements.yml both in the PAH as in our config -> same error.
3) Via web interface we never had any trouble to upload collections which are already part of another repository. They just appear at community repo and at published repo (which is not too clean, but never failed).

Actually infra.controller_configuration does not fail, although it is already in the validated collections repository, just in an older version.


> With your `debeka_automation.infrastructure` collection this seems a slightly different issue. I think given the logs you've provided there isn't enough info to diagnose. I'd suggest taking a look in the tasks section of the UI when running this as this should return you a log of what happened during the attempted upload. Could you post this in please so we can diagnose further.

There is not failed task there.
When I run the Playbook, there appear four tasks, all four in status "completed" before the Ansible-Run is failing: twice "pulpcore.app.tasks.base.general_update" and twice "galaxy_ng.app.tasks.namespaces._create_pulp_namespace"


Is there anywhere else where I could have a look?

The error message seems to say (in my words) "You are not allowed to send a POST request to /api/galaxy/v3/artifacts/collections/" or something like that.

Yours
  David

**Comment by dbk-rabel** [comment]
> Actually infra.controller_configuration does not fail, although it is already in the validated collections repository, just in an older version.


While thinking about this a little bit longer, I realized that the namespaces community and debeka_automation are created, but empty. And that community.general and debeka_automation.infrastructure are not present.

Could there be a problem when
a) uploading collections to a namespace that has no collections in it yet or
b) uploading collections which are not present in other versions on the PAH yet?

Or even something else like a problem with the namespaces that infra.ah_configuration creates?

Yours
  David

**Comment by Tompage1994** [comment]
@dbk-rabel sorry for delay in getting back to you.

From memory, yes you do need to have first created the namespace before trying to upload the collection. I don't believe the namespace will automatically get created. I think, though, that if the namespace does exist you should be able to upload a completely new collection.

Hope this answers your question

**Comment by dbk-rabel** [comment]
> Hope this answers your question

Actually not. Sorry :D 

We do create the namespace before trying to push the collection. The question was if there could be a problem if the namespace is still empty. So if the first collection per namespace has to be pushed in a different way.

**Comment by Tompage1994** [comment]
Ok, then this shouldn't be an issue. To verify I've run the following without failure:

```
---
- name: Test playbook
  hosts: localhost
  vars:
  my_collection:
      name: galaxy
      namespace: new_ns
      path: new_ns-galaxy-1.1.1-devel.tar.gz


  tasks:
    - name: Create NS
      infra.ah_configuration.ah_namespace:
        name: "{{ my_collection.namespace }}"
        state: present

    - name: Create collection
      infra.ah_configuration.ah_collection:
        name: "{{ my_collection.name }}"
        namespace: "{{ my_collection.namespace }}"
        path: "{{ my_collection.path }}"
        state: present
        auto_approve: true  # default
        wait: true
```

So now I can't replicate your issue. All I can think is that something is going wrong on the approval step. Could you verify the following:
- Are the collections sitting waiting for approval? https://<ah_url>/ui/approval-dashboard/
- Can you check if the play works when setting `auto_approve` to false for each collection, does the collection go into approval state then?

**Comment by dbk-rabel** [comment]
I'm not able to reproduce the exact error. We had to fix some things manually and now it is difficult to break it properly without re-installing AAP2. (But I will do that, if it will be necessary to track down the error :) )

For the moment I am receiving a HTTP403 on a GET request, which indicates that the upload works fine and then there is a problem with the auto-approval.

But the interesting question is "Why?".


Instead of your playbook I used the `infra.ah_configuration.collection` role:

```
---
- name: Test playbook
  hosts: localhost

  tasks:
    - name: Include role infra.ah_configuration.collection
      ansible.builtin.include_role:
        name: infra.ah_configuration.collection
```

This does work when only having one collection in `ah_collections`, but fails when I have two or more.

So I took a look at the API logs and saw that there is a second POST request to `/api/galaxy/v3/auth/token/`

Maybe this is because the role is trying to handle both collections in parallel, therefore tries to create a token and then... I don't know. Maybe this invalidates the first token? Or the the module confuses both tokens?

Can you reproduce this behaviour?


One collection:

```
<ip address> - admin [17/Apr/2024:18:45:48 +0200] "POST /api/galaxy/v3/auth/token/ HTTP/1.1" 200 52 "-" "Python-urllib/3.6" "-"
<ip address> - - [17/Apr/2024:18:45:48 +0200] "GET /api/galaxy/v3/collections/debeka_automation/infrastructure/versions/2.0.3/ HTTP/1.1" 302 0 "-" "Python-urllib/3.6" "-"
<ip address> - - [17/Apr/2024:18:45:48 +0200] "GET /api/galaxy/v3/plugin/ansible/content/published/collections/index/debeka_automation/infrastructure/versions/2.0.3/ HTTP/1.1" 404 69 "-" "Python-urllib/3.6" "-"
<ip address> - - [17/Apr/2024:18:45:48 +0200] "POST /api/galaxy/v3/artifacts/collections/ HTTP/1.1" 202 114 "-" "Python-urllib/3.6" "-"
<ip address> - - [17/Apr/2024:18:45:48 +0200] "GET /api/galaxy/content/staging/v3/plugin/ansible/imports/collections/018eecf3-4142-70c1-bbc7-115f7ef90c6a/ HTTP/1.1" 200 238 "-" "Python-urllib/3.6" "-"
<ip address> - - [17/Apr/2024:18:45:49 +0200] "GET /api/galaxy/content/staging/v3/plugin/ansible/imports/collections/018eecf3-4142-70c1-bbc7-115f7ef90c6a/ HTTP/1.1" 200 1224 "-" "Python-urllib/3.6" "-"
<ip address> - - [17/Apr/2024:18:45:51 +0200] "GET /api/galaxy/content/staging/v3/plugin/ansible/imports/collections/018eecf3-4142-70c1-bbc7-115f7ef90c6a/ HTTP/1.1" 200 1224 "-" "Python-urllib/3.6" "-"
<ip address> - - [17/Apr/2024:18:45:52 +0200] "GET /api/galaxy/content/staging/v3/plugin/ansible/imports/collections/018eecf3-4142-70c1-bbc7-115f7ef90c6a/ HTTP/1.1" 200 1690 "-" "Python-urllib/3.6" "-"
<ip address> - - [17/Apr/2024:18:45:54 +0200] "POST /api/galaxy/v3/collections/debeka_automation/infrastructure/versions/2.0.3/move/staging/published/ HTTP/1.1" 202 157 "-" "Python-urllib/3.6" "-"
```


Two collections:

```
<ip address> - admin [17/Apr/2024:18:17:17 +0200] "POST /api/galaxy/v3/auth/token/ HTTP/1.1" 200 52 "-" "Python-urllib/3.6" "-"
<ip address> - - [17/Apr/2024:18:17:17 +0200] "GET /api/galaxy/v3/collections/debeka_automation/infrastructure/versions/2.0.3/ HTTP/1.1" 302 0 "-" "Python-urllib/3.6" "-"
<ip address> - - [17/Apr/2024:18:17:17 +0200] "GET /api/galaxy/v3/plugin/ansible/content/published/collections/index/debeka_automation/infrastructure/versions/2.0.3/ HTTP/1.1" 404 69 "-" "Python-urllib/3.6" "-"
<ip address> - - [17/Apr/2024:18:17:17 +0200] "POST /api/galaxy/v3/artifacts/collections/ HTTP/1.1" 202 114 "-" "Python-urllib/3.6" "-"
<ip address> - - [17/Apr/2024:18:17:17 +0200] "GET /api/galaxy/content/staging/v3/plugin/ansible/imports/collections/018eecd9-25e3-7dc2-bf4a-89d555451d5c/ HTTP/1.1" 200 238 "-" "Python-urllib/3.6" "-"
<ip address> - admin [17/Apr/2024:18:17:18 +0200] "POST /api/galaxy/v3/auth/token/ HTTP/1.1" 200 52 "-" "Python-urllib/3.6" "-"
<ip address> - - [17/Apr/2024:18:17:18 +0200] "GET /api/galaxy/v3/collections/infra/controller_configuration/versions/2.7.1/ HTTP/1.1" 302 0 "-" "Python-urllib/3.6" "-"
<ip address> - - [17/Apr/2024:18:17:18 +0200] "GET /api/galaxy/v3/plugin/ansible/content/published/collections/index/infra/controller_configuration/versions/2.7.1/ HTTP/1.1" 200 158847 "-" "Python-urllib/3.6" "-"
<ip address> - - [17/Apr/2024:18:17:18 +0200] "GET /api/galaxy/content/staging/v3/plugin/ansible/imports/collections/018eecd9-25e3-7dc2-bf4a-89d555451d5c/ HTTP/1.1" 403 134 "-" "Python-urllib/3.6" "-"
```

**Comment by Tompage1994** [comment]
Ah, ok I've got it now. and I've replicated.

The issue very much does seem to be that the token gets invalidated because a new one is created. I'll do some exploring to see what the issue is.

`So I took a look at the API logs and saw that there is a second POST request to /api/galaxy/v3/auth/token/`

**Comment by Tompage1994** [comment]
As a workaround, you could make use of the ah_token module which will set you a token which persists for use of the roles:

```
    - name: Get ah_token
      infra.ah_configuration.ah_token:
        ah_username: "{{ ah_username }}"
        ah_password: "{{ ah_password }}"
        ah_hostname: "{{ ah_host }}"
```

I've validated this should work fine. I will dive now into trying to fix the issue here though

**Comment by dbk-rabel** [comment]
Thanks a lot! I will try that.
